source/app.d

```diff
@@ -51,13 +52,22 @@ void main(string[] args)
 	if (!CLI!DlingsCmdlArgs.parseArgs(cmdlArgs, args[1 .. $]))
 		return;
 
+
+	if (!exists("exercises.toml"))
+	{
+		writeln("file 'exercises.toml' is missing!");
+		return;
+	}
+
+	immutable Exercise[] exercises = "exercises.toml".exercisesFromTOML();
```

Comment:
Since you are parsing UTF-8 text, you can handle a custom message like you done above for `.toml` files that are not UTF-8 compliant.

Reply:
`readText` throws an exception on bad UTF-8 reading. You can handle it to pretty print an error message rather than the default exception message + stacktrace, like you have done above.

Reply:
where is that information stated?

Reply:
https://dlang.org/phobos/std_file.html#readText

Reply:
Does the current change look alright?